sqld/src/hrana/conn.rs

```diff
@@ -172,25 +203,47 @@ async fn handle_request_msg(
         return Ok(false)
     };
 
-    match session::handle_request(&conn.server, session, request).await {
-        Ok(response) => {
-            send_msg(
-                conn,
-                &proto::ServerMsg::ResponseOk {
-                    request_id,
-                    response,
-                },
-            )
-            .await?;
-            Ok(true)
-        }
-        Err(err) => match downcast_error(err) {
-            Ok(error) => {
-                send_msg(conn, &proto::ServerMsg::ResponseError { request_id, error }).await?;
-                Ok(true)
+    let response_rx = session::handle_request(&conn.server, session, &mut conn.join_set, request)
+        .await
+        .unwrap_or_else(|err| {
+            // we got an error immediately, but let's treat it as a special case of the general
+            // flow
+            let (tx, rx) = oneshot::channel();
+            tx.send(Err(err)).unwrap();
+            rx
+        });
+
+    conn.responses.push(ResponseFuture {
+        request_id,
+        response_rx: response_rx.fuse(),
+    });
+    Ok(true)
+}
+
+impl Future for ResponseFuture {
+    type Output = Result<proto::ServerMsg>;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match ready!(Pin::new(&mut self.response_rx).poll(cx)) {
+            Ok(Ok(response)) => Poll::Ready(Ok(proto::ServerMsg::ResponseOk {
+                request_id: self.request_id,
+                response,
+            })),
+            Ok(Err(err)) => match downcast_error(err) {
+                Ok(error) => Poll::Ready(Ok(proto::ServerMsg::ResponseError {
+                    request_id: self.request_id,
+                    error,
+                })),
+                Err(err) => Poll::Ready(Err(err)),
+            },
+            Err(_recv_err) => {
+                // do not propagate this error, because the error that caused the receiver to drop
+                // is very likely propagating from another task at this moment, and we don't want
+                // to hide it.
+                // this is also the reason why we need to use `Fuse` in self.response_rx
+                tracing::warn!("Response sender was dropped");
+                Poll::Pending
```

Comment:
Yes, that's the assumption. I feel a bit uneasy about this, but I have very bad experience with programs crashing due to "channel closed" errors, without giving a clue about the true source of the error.

The worst thing that can happen is if we drop the sender without raising an error, which would cause the futures to accumulate. The warning is an attempt to mitigate this, because the logs will show that something is wrong.